### DISCUSSION TITLE
Remapping DDC brightness control to contrast on external display?

### DISCUSSION THREAD

**Discussion by renaudguerin** [MonitorControl/MonitorControl#764] [General]
### Before opening the issue, have you...?

- [X] Searched for existing issues

### Is your feature request related to a problem? Please describe

My work setup is an Intel MBP -> CalDigit Thunderbolt hub -> USB-C to Displayport cable -> BenQ BL3201PT

This monitor doesn't respond to brightness controls at all with MonitorControl 3.1.1
It does respond to volume and contrast though, and the contrast menu slider works fine.
This means the brightness keys do nothing, and I have to reach for the menu instead.

### Describe the solution you'd like

A preference checkbox to enable permanent remapping of Brightness keyboard controls to Contrast instead.
Or (probably cleaner) don't hardcode the mapping of keys -> feature name -> ID at all but make it configurable, with sane defaults out of the box.

### Describe alternatives you've considered

As I was writing this I've discovered that 4.0.0 introduces a keyboard shortcut for contrast (Ctrl-Add-Cmd + brightness key)

This is an improvement, but it mostly benefits people who are already able to control brightness using the unmodified keys. In my case, the brightness keys are useless without this complex modifier, so I would much prefer direct access.

I appreciate the concerns around not cluttering the preferences UI, but IMHO some of the current settings are even more niche than this one (e.g. "Further lower brightness via software dimming")

Maybe a "batteries included but optional" approach would work here ? i.e. don't hardcode brightness/contrast/volume controls per se, make the brightness keys-> feature name ->IDs mapping configurable for advanced users, and ship with sane defaults that work for most people.

**Comment by renaudguerin** [comment]
Ok sorry, it looks like I missed this in the 4.0.0 release notes which seems relevant  :
 _Added comma separated list for control code override to enable edge cases like controlling Brightness and Contrast at the same time (use VCP list entry 10, 12 for that)_
 
I've just upgraded to 4.0.0 but unfortunately neither setting `10, 12` or even just `12` in the VCP list entry for "Brightness" makes a difference : the brightness OSD shows up when using the keys but contrast isn't affected.

**Comment by waydabber** [comment]
Does the contrast setting work by itself? Entering 12 to the VCP list should remap Brightness to Contrast. Can you show me a screenshot of your Displays tab? Thanks!

**Comment by renaudguerin** [comment]
Thanks for the reply.
Yes both the contrast and volume settings work fine by themselves.
But nothing from brightness keys/sliders, even with VCP set to 12 for Brightness.
<img width="708" alt="Screenshot 2021-11-04 at 01 13 55" src="https://user-images.githubusercontent.com/1047312/140240906-48eb8c58-5054-44fd-8f95-9ce591c17630.png">

**Comment by renaudguerin** [comment]
If that helps, I've played around with `ddcctl` and it does manage to query + change contrast :
Query :
```
$ ddcctl -d 1 -c ?
D: CGDisplay 47C59827-56AA-8843-E35D-E9F84FDB1D32 dispID(#1952469573) (1920x1080 0°) HiDPI
I: found 1 external display
I: polling EDID for #1 (ID 1952469573 => IOService:/AppleACPIPlatformExpert/PCI0@0/AppleACPIPCI/IGPU@2/AppleIntelFramebuffer@1)
I: got edid.serial: R2F05134SL0
I: got edid.name: BenQ LCD
D: action: c: ?
D: querying VCP control: #18 =?
D: Tries required to get data: 3
I: VCP control #18 (0x12) = current: 76, max: 100
```

Set :
```
$ ddcctl -d 1 -c 50
D: CGDisplay 47C59827-56AA-8843-E35D-E9F84FDB1D32 dispID(#1952469573) (1920x1080 0°) HiDPI
I: found 1 external display
I: polling EDID for #1 (ID 1952469573 => IOService:/AppleACPIPlatformExpert/PCI0@0/AppleACPIPCI/IGPU@2/AppleIntelFramebuffer@1)
I: got edid.serial: R2F05134SL0
I: got edid.name: BenQ LCD
D: action: c: 50
D: setting VCP control #18 => 50
```

But again, the Contrast slider in MonitorControl works as well, so I suspect there's just something wrong with that new "VCP list" feature.

**Comment by renaudguerin** [comment]
And interestingly, setting brightness with `ddcctl` (VCP 0x10) then querying it immediately after returns the value you've just set. So it is stored somewhere, but just has no effect on the monitor whatsoever.

**Comment by renaudguerin** [comment]
Using `ddcctl` to query the brightness value while using the brightness keys/slider in MonitorControl with "VCP list" = 12, I can confirm that they're still only controlling VCP 0x10, not 0x12.

**Comment by waydabber** [comment]
Interesting. If you enter 62, does it control volume instead? Does the software dimming part of combined brightness work?

**Reply by renaudguerin** [reply]
> Interesting. If you enter 62, does it control volume instead? Does the software dimming part of combined brightness work?

The software dimming part of "combined brightness" kind of works (kicks in under 50% only). I was gonna say it's not usable for me because it doesn't play nice with f.lux, but enabling "avoid gamma table manipulation" fixes that.

Disabling "Hardware DDC control" + Enabling "Avoid Gamma manipulation" would actually be a good enough workaround for brightness (kicks in not just under 50%, plays nice with f.lux), unfortunately that also disables volume control 😞

And no entering 62 doesn't work either, even though the volume keys and slider work perfectly (with hardware DDC control enabled ofc)

I would say the bug probably has nothing to do with the DDC code (the contrast slider is working fine). It really feels as if the UI isn't passing on the value at all.

I'm on Monterey and this is an Intel MBP, if that matters. I also have an M1 Macbook Air, I will test with it later

**Reply by waydabber** [reply]
Ohh, so this is Intel. You are right, I just checked the code, on the Intel codepath I forgot to update a variable name hehe. :) I was testing on M1... 

I'll fix this in v4.0.1, thanks!

**Reply by waydabber** [reply]
Fixed already.

**Reply by waydabber** [reply]
Sorry, I see you mentioned Intel from the get go, I just somehow did not realise it all along (even though you mentioned using ddcctl etc)...

**Reply by renaudguerin** [reply]
Awesome thanks ! Looking forward to it.

Just in case I end up preferring the purely software brightness control (the one that happens when both Hardware DDC control and "combined brightness" are off, and kicks in not only under <50), is there any way to get that for brightness but keep hardware DDC for volume / contrast ?

**Reply by waydabber** [reply]
Well, no. But you can get very close to it by enabling "combined" and then moving the "combined dimming switchover point" all the way to the right.

**Reply by renaudguerin** [reply]
Perfect thanks!
